File: DataOutput/clsSpectrumDataWriter.vb
Imports MASIC.clsMASIC

Namespace DataOutput

    Public Class clsSpectrumDataWriter
        Inherits clsMasicEventNotifier

#Region "Classwide variables"
        Private ReadOnly mBPIWriter As clsBPIWriter
        Private ReadOnly mOptions As clsMASICOptions
#End Region

        ''' <summary>
        ''' Constructor
        ''' </summary>
        Public Sub New(bpiWriter As clsBPIWriter, masicOptions As clsMASICOptions)
            mBPIWriter = bpiWriter
            mOptions = masicOptions
        End Sub

        Public Function ExportRawDataToDisk(
          scanList As clsScanList,
          objSpectraCache As clsSpectraCache,
          inputFileName As String,
          outputDirectoryPath As String) As Boolean

            Dim outputFilePath = "??"

            Try
                Dim dataWriter As StreamWriter
                Dim scanInfoWriter As StreamWriter

                Select Case mOptions.RawDataExportOptions.FileFormat
                    Case clsRawDataExportOptions.eExportRawDataFileFormatConstants.PEKFile
                        outputFilePath = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.PEKFile)
                        dataWriter = New StreamWriter(outputFilePath)
                        scanInfoWriter = Nothing

                    Case clsRawDataExportOptions.eExportRawDataFileFormatConstants.CSVFile
                        outputFilePath = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.DeconToolsIsosFile)

                        Dim outputFilePath2 = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.DeconToolsScansFile)

                        dataWriter = New StreamWriter(outputFilePath)
                        scanInfoWriter = New StreamWriter(outputFilePath2)

                        ' Write the file headers
                        mBPIWriter.WriteDecon2LSIsosFileHeaders(dataWriter)
                        mBPIWriter.WriteDecon2LSScanFileHeaders(scanInfoWriter)

                    Case Else
                        ' Unknown format
                        ReportError("Unknown raw data file format: " & mOptions.RawDataExportOptions.FileFormat.ToString())
                        Return False
                End Select

                Dim spectrumExportCount = 0

                If Not mOptions.RawDataExportOptions.IncludeMSMS AndAlso mOptions.RawDataExportOptions.RenumberScans Then
                    mOptions.RawDataExportOptions.RenumberScans = True
                Else
                    mOptions.RawDataExportOptions.RenumberScans = False
                End If

                UpdateProgress(0, "Exporting raw data")

                For masterOrderIndex = 0 To scanList.MasterScanOrderCount - 1
                    Dim scanPointer = scanList.MasterScanOrder(masterOrderIndex).ScanIndexPointer
                    If scanList.MasterScanOrder(masterOrderIndex).ScanType = clsScanList.eScanTypeConstants.SurveyScan Then
                        SaveRawDataToDiskWork(dataWriter, scanInfoWriter, scanList.SurveyScans(scanPointer), objSpectraCache, inputFileName, False, spectrumExportCount)
                    Else
                        If mOptions.RawDataExportOptions.IncludeMSMS OrElse
                          Not scanList.FragScans(scanPointer).MRMScanType = ThermoRawFileReader.MRMScanTypeConstants.NotMRM Then
                            ' Either we're writing out MS/MS data or this is an MRM scan
                            SaveRawDataToDiskWork(dataWriter, scanInfoWriter, scanList.FragScans(scanPointer), objSpectraCache, inputFileName, True, spectrumExportCount)
                        End If
                    End If

                    If scanList.MasterScanOrderCount > 1 Then
                        UpdateProgress(CShort(masterOrderIndex / (scanList.MasterScanOrderCount - 1) * 100))
                    Else
                        UpdateProgress(0)
                    End If

                    UpdateCacheStats(objSpectraCache)

                    If mOptions.AbortProcessing Then
                        Exit For
                    End If
                Next

                If Not dataWriter Is Nothing Then dataWriter.Close()
                If Not scanInfoWriter Is Nothing Then scanInfoWriter.Close()

                Return True

            Catch ex As Exception
                ReportError("Error writing the raw spectra data to: " & outputFilePath, ex, eMasicErrorCodes.OutputFileWriteError)
                Return False
            End Try

        End Function


        Private Sub SaveCSVFilesToDiskWork(
          dataWriter As StreamWriter,
          scanInfoWriter As StreamWriter,
          currentScan As clsScanInfo,
          objSpectraCache As clsSpectraCache,
          fragmentationScan As Boolean,
          ByRef spectrumExportCount As Integer)

            Dim poolIndex As Integer
            Dim scanNumber As Integer
            Dim baselineNoiseLevel As Double

            If Not objSpectraCache.ValidateSpectrumInPool(currentScan.ScanNumber, poolIndex) Then
                SetLocalErrorCode(eMasicErrorCodes.ErrorUncachingSpectrum)
                Exit Sub
            End If

            spectrumExportCount += 1

            With currentScan
                ' First, write an entry to the "_scans.csv" file

                If mOptions.RawDataExportOptions.RenumberScans Then
                    scanNumber = spectrumExportCount
                Else
                    scanNumber = .ScanNumber
                End If

                Dim msLevel As Integer
                If fragmentationScan Then
                    msLevel = .FragScanInfo.MSLevel
                Else
                    msLevel = 1
                End If

                Dim numIsotopicSignatures = 0
                Dim numPeaks = objSpectraCache.SpectraPool(poolIndex).IonCount

                baselineNoiseLevel = .BaselineNoiseStats.NoiseLevel
                If baselineNoiseLevel < 1 Then baselineNoiseLevel = 1

                ' Old Column Order:
                ''lineOut = scanNumber.ToString() & "," &
                ''             .ScanTime.ToString("0.0000") &
                ''             msLevel & "," &
                ''             numIsotopicSignatures & "," &
                ''             numPeaks & "," &
                ''             .TotalIonIntensity.ToString() & "," &
                ''             .BasePeakIonMZ & "," &
                ''             .BasePeakIonIntensity & "," &
                ''             timeDomainIntensity & "," &
                ''             peakIntensityThreshold & "," &
                ''             peptideIntensityThreshold

                mBPIWriter.WriteDecon2LSScanFileEntry(scanInfoWriter, currentScan, scanNumber, msLevel, numPeaks, numIsotopicSignatures)
            End With


            With objSpectraCache.SpectraPool(poolIndex)
                ' Now write an entry to the "_isos.csv" file

                If .IonCount > 0 Then
                    ' Populate intensities and pointerArray()

                    Dim intensities() As Double
                    Dim pointerArray() As Integer

                    ReDim intensities(.IonCount - 1)
                    ReDim pointerArray(.IonCount - 1)
                    For ionIndex = 0 To .IonCount - 1
                        intensities(ionIndex) = .IonsIntensity(ionIndex)
                        pointerArray(ionIndex) = ionIndex
                    Next

                    ' Sort pointerArray() based on the intensities in intensities
                    Array.Sort(intensities, pointerArray)

                    Dim startIndex As Integer
                    If mOptions.RawDataExportOptions.MaxIonCountPerScan > 0 Then
                        ' Possibly limit the number of ions to maxIonCount
                        startIndex = .IonCount - mOptions.RawDataExportOptions.MaxIonCountPerScan
                        If startIndex < 0 Then startIndex = 0
                    Else
                        startIndex = 0
                    End If

                    ' Define the minimum data point intensity value
                    Dim minimumIntensityCurrentScan = .IonsIntensity(pointerArray(startIndex))

                    ' Update the minimum intensity if a higher minimum intensity is defined in .IntensityMinimum
                    minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, mOptions.RawDataExportOptions.IntensityMinimum)

                    ' If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio is > 0, then possibly update minimumIntensityCurrentScan
                    If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio > 0 Then
                        minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, currentScan.BaselineNoiseStats.NoiseLevel * mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio)
                    End If

                    For ionIndex = 0 To .IonCount - 1
                        If .IonsIntensity(ionIndex) >= minimumIntensityCurrentScan Then

                            Dim charge = 1
                            Dim isoFit = 0
                            Dim mass = clsUtilities.ConvoluteMass(.IonsMZ(ionIndex), 1, 0)
                            Dim peakFWHM = 0
                            Dim signalToNoise = .IonsIntensity(ionIndex) / baselineNoiseLevel
                            Dim monoisotopicAbu = -10
                            Dim monoPlus2Abu = -10

                            mBPIWriter.WriteDecon2LSIsosFileEntry(
                              dataWriter, scanNumber, charge,
                              .IonsIntensity(ionIndex), .IonsMZ(ionIndex),
                              isoFit, mass, mass, mass,
                              peakFWHM, signalToNoise, monoisotopicAbu, monoPlus2Abu)

                        End If
                    Next
                End If

            End With

        End Sub

        Private Sub SavePEKFileToDiskWork(
          writer As TextWriter,
          currentScan As clsScanInfo,
          objSpectraCache As clsSpectraCache,
          inputFileName As String,
          fragmentationScan As Boolean,
          ByRef spectrumExportCount As Integer)

            Dim poolIndex As Integer
            Dim exportCount = 0

            If Not objSpectraCache.ValidateSpectrumInPool(currentScan.ScanNumber, poolIndex) Then
                SetLocalErrorCode(eMasicErrorCodes.ErrorUncachingSpectrum)
                Exit Sub
            End If

            spectrumExportCount += 1

            With currentScan

                writer.WriteLine("Time domain signal level:" & ControlChars.Tab & .BasePeakIonIntensity.ToString())          ' Store the base peak ion intensity as the time domain signal level value

                writer.WriteLine("MASIC " & mOptions.MASICVersion)                     ' Software version
                Dim dataLine = "MS/MS-based PEK file"
                If mOptions.RawDataExportOptions.IncludeMSMS Then
                    dataLine &= " (includes both survey scans and fragmentation spectra)"
                Else
                    dataLine &= " (includes only survey scans)"
                End If
                writer.WriteLine(dataLine)

                Dim scanNumber As Integer
                If mOptions.RawDataExportOptions.RenumberScans Then
                    scanNumber = spectrumExportCount
                Else
                    scanNumber = .ScanNumber
                End If
                dataLine = "Filename: " & inputFileName & "." & scanNumber.ToString("00000")
                writer.WriteLine(dataLine)

                If fragmentationScan Then
                    writer.WriteLine("ScanType: Fragmentation Scan")
                Else
                    writer.WriteLine("ScanType: Survey Scan")
                End If

                writer.WriteLine("Charge state mass transform results:")
                writer.WriteLine("First CS,    Number of CS,   Abundance,   Mass,   Standard deviation")
            End With

            With objSpectraCache.SpectraPool(poolIndex)

                If .IonCount > 0 Then
                    ' Populate intensities and pointerArray()
                    Dim intensities() As Double
                    Dim pointerArray() As Integer

                    ReDim intensities(.IonCount - 1)
                    ReDim pointerArray(.IonCount - 1)
                    For ionIndex = 0 To .IonCount - 1
                        intensities(ionIndex) = .IonsIntensity(ionIndex)
                        pointerArray(ionIndex) = ionIndex
                    Next

                    ' Sort pointerArray() based on the intensities in intensities
                    Array.Sort(intensities, pointerArray)

                    Dim startIndex As Integer

                    If mOptions.RawDataExportOptions.MaxIonCountPerScan > 0 Then
                        ' Possibly limit the number of ions to maxIonCount
                        startIndex = .IonCount - mOptions.RawDataExportOptions.MaxIonCountPerScan
                        If startIndex < 0 Then startIndex = 0
                    Else
                        startIndex = 0
                    End If

                    ' Define the minimum data point intensity value
                    Dim minimumIntensityCurrentScan = .IonsIntensity(pointerArray(startIndex))

                    ' Update the minimum intensity if a higher minimum intensity is defined in .IntensityMinimum
                    minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, mOptions.RawDataExportOptions.IntensityMinimum)

                    ' If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio is > 0, then possibly update minimumIntensityCurrentScan
                    If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio > 0 Then
                        minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, currentScan.BaselineNoiseStats.NoiseLevel * mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio)
                    End If

                    exportCount = 0
                    For ionIndex = 0 To .IonCount - 1
                        If .IonsIntensity(ionIndex) >= minimumIntensityCurrentScan Then
                            Dim dataLine =
                                "1" & ControlChars.Tab &
                                "1" & ControlChars.Tab &
                                .IonsIntensity(ionIndex) & ControlChars.Tab &
                                .IonsMZ(ionIndex) & ControlChars.Tab &
                                "0"

                            writer.WriteLine(dataLine)
                            exportCount += 1
                        End If
                    Next
                End If

                writer.WriteLine("Number of peaks in spectrum = " & .IonCount.ToString())
                writer.WriteLine("Number of isotopic distributions detected = " & exportCount.ToString())
                writer.WriteLine()

            End With

        End Sub

        Private Sub SaveRawDataToDiskWork(
          dataWriter As StreamWriter,
          scanInfoWriter As StreamWriter,
          currentScan As clsScanInfo,
          objSpectraCache As clsSpectraCache,
          inputFileName As String,
          fragmentationScan As Boolean,
          ByRef spectrumExportCount As Integer)

            Select Case mOptions.RawDataExportOptions.FileFormat
                Case clsRawDataExportOptions.eExportRawDataFileFormatConstants.PEKFile
                    SavePEKFileToDiskWork(dataWriter, currentScan, objSpectraCache, inputFileName, fragmentationScan, spectrumExportCount)
                Case clsRawDataExportOptions.eExportRawDataFileFormatConstants.CSVFile
                    SaveCSVFilesToDiskWork(dataWriter, scanInfoWriter, currentScan, objSpectraCache, fragmentationScan, spectrumExportCount)
                Case Else
                    ' Unknown format
                    ' This code should never be reached
            End Select
        End Sub

    End Class

End Namespace

File: DataOutput/clsSpectrumDataWriter.cs
using System;
using System.IO;

namespace MASIC.DataOutput
{
    public class clsSpectrumDataWriter : clsMasicEventNotifier
    {
        #region "Classwide variables"
        private readonly clsBPIWriter mBPIWriter;
        private readonly clsMASICOptions mOptions;
        #endregion

        /// <summary>
        /// Constructor
        /// </summary>
        public clsSpectrumDataWriter(clsBPIWriter bpiWriter, clsMASICOptions masicOptions)
        {
            mBPIWriter = bpiWriter;
            mOptions = masicOptions;
        }

        public bool ExportRawDataToDisk(
            clsScanList scanList,
            clsSpectraCache spectraCache,
            string inputFileName,
            string outputDirectoryPath)
        {
            var outputFilePath = "??";

            try
            {
                StreamWriter dataWriter;
                StreamWriter scanInfoWriter;

                switch (mOptions.RawDataExportOptions.FileFormat)
                {
                    case clsRawDataExportOptions.eExportRawDataFileFormatConstants.PEKFile:
                        outputFilePath = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.PEKFile);
                        dataWriter = new StreamWriter(outputFilePath);
                        scanInfoWriter = null;
                        break;
                    case clsRawDataExportOptions.eExportRawDataFileFormatConstants.CSVFile:
                        outputFilePath = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.DeconToolsIsosFile);

                        var outputFilePath2 = clsDataOutput.ConstructOutputFilePath(inputFileName, outputDirectoryPath, clsDataOutput.eOutputFileTypeConstants.DeconToolsScansFile);

                        dataWriter = new StreamWriter(outputFilePath);
                        scanInfoWriter = new StreamWriter(outputFilePath2);

                        // Write the file headers
                        mBPIWriter.WriteDecon2LSIsosFileHeaders(dataWriter);
                        mBPIWriter.WriteDecon2LSScanFileHeaders(scanInfoWriter);
                        break;

                    default:
                        // Unknown format
                        ReportError("Unknown raw data file format: " + mOptions.RawDataExportOptions.FileFormat.ToString());
                        return false;
                }

                var spectrumExportCount = 0;

                if (!mOptions.RawDataExportOptions.IncludeMSMS && mOptions.RawDataExportOptions.RenumberScans)
                {
                    mOptions.RawDataExportOptions.RenumberScans = true;
                }
                else
                {
                    mOptions.RawDataExportOptions.RenumberScans = false;
                }

                UpdateProgress(0, "Exporting raw data");

                for (var masterOrderIndex = 0; masterOrderIndex < scanList.MasterScanOrderCount; masterOrderIndex++)
                {
                    var scanPointer = scanList.MasterScanOrder[masterOrderIndex].ScanIndexPointer;
                    if (scanList.MasterScanOrder[masterOrderIndex].ScanType == clsScanList.eScanTypeConstants.SurveyScan)
                    {
                        SaveRawDataToDiskWork(dataWriter, scanInfoWriter, scanList.SurveyScans[scanPointer], spectraCache, inputFileName, false, ref spectrumExportCount);
                    }
                    else if (mOptions.RawDataExportOptions.IncludeMSMS ||
                        scanList.FragScans[scanPointer].MRMScanType != ThermoRawFileReader.MRMScanTypeConstants.NotMRM)
                    {
                        // Either we're writing out MS/MS data or this is an MRM scan
                        SaveRawDataToDiskWork(dataWriter, scanInfoWriter, scanList.FragScans[scanPointer], spectraCache, inputFileName, true, ref spectrumExportCount);
                    }

                    if (scanList.MasterScanOrderCount > 1)
                    {
                        UpdateProgress((short)(masterOrderIndex / (double)(scanList.MasterScanOrderCount - 1) * 100));
                    }
                    else
                    {
                        UpdateProgress(0);
                    }

                    UpdateCacheStats(spectraCache);

                    if (mOptions.AbortProcessing)
                    {
                        break;
                    }
                }

                dataWriter.Close();
                scanInfoWriter?.Close();

                return true;
            }
            catch (Exception ex)
            {
                ReportError("Error writing the raw spectra data to: " + outputFilePath, ex, clsMASIC.eMasicErrorCodes.OutputFileWriteError);
                return false;
            }
        }

        private void SaveCSVFilesToDiskWork(
            StreamWriter dataWriter,
            StreamWriter scanInfoWriter,
            clsScanInfo currentScan,
            clsSpectraCache spectraCache,
            bool fragmentationScan,
            ref int spectrumExportCount)
        {
            int scanNumber;

            if (!spectraCache.GetSpectrum(currentScan.ScanNumber, out var spectrum, true))
            {
                SetLocalErrorCode(clsMASIC.eMasicErrorCodes.ErrorUncachingSpectrum);
                return;
            }

            spectrumExportCount += 1;

            // First, write an entry to the "_scans.csv" file

            if (mOptions.RawDataExportOptions.RenumberScans)
            {
                scanNumber = spectrumExportCount;
            }
            else
            {
                scanNumber = currentScan.ScanNumber;
            }

            int msLevel;
            if (fragmentationScan)
            {
                msLevel = currentScan.FragScanInfo.MSLevel;
            }
            else
            {
                msLevel = 1;
            }

            var numIsotopicSignatures = 0;
            var numPeaks = spectrum.IonCount;

            var baselineNoiseLevel = currentScan.BaselineNoiseStats.NoiseLevel;
            if (baselineNoiseLevel < 1)
                baselineNoiseLevel = 1;

            mBPIWriter.WriteDecon2LSScanFileEntry(scanInfoWriter, currentScan, scanNumber, msLevel, numPeaks, numIsotopicSignatures);

            // Now write an entry to the "_isos.csv" file

            if (spectrum.IonCount > 0)
            {
                // Populate intensities and pointerArray()

                var intensities = new double[spectrum.IonCount];
                var pointerArray = new int[spectrum.IonCount];
                for (var ionIndex = 0; ionIndex < spectrum.IonCount; ionIndex++)
                {
                    intensities[ionIndex] = spectrum.IonsIntensity[ionIndex];
                    pointerArray[ionIndex] = ionIndex;
                }

                // Sort pointerArray() based on the intensities in intensities
                Array.Sort(intensities, pointerArray);

                int startIndex;
                if (mOptions.RawDataExportOptions.MaxIonCountPerScan > 0)
                {
                    // Possibly limit the number of ions to maxIonCount
                    startIndex = spectrum.IonCount - mOptions.RawDataExportOptions.MaxIonCountPerScan;
                    if (startIndex < 0)
                        startIndex = 0;
                }
                else
                {
                    startIndex = 0;
                }

                // Define the minimum data point intensity value
                var minimumIntensityCurrentScan = spectrum.IonsIntensity[pointerArray[startIndex]];

                // Update the minimum intensity if a higher minimum intensity is defined in .IntensityMinimum
                minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, mOptions.RawDataExportOptions.IntensityMinimum);

                // If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio is > 0, then possibly update minimumIntensityCurrentScan
                if (mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio > 0)
                {
                    minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, currentScan.BaselineNoiseStats.NoiseLevel * mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio);
                }

                for (var ionIndex = 0; ionIndex < spectrum.IonCount; ionIndex++)
                {
                    if (spectrum.IonsIntensity[ionIndex] >= minimumIntensityCurrentScan)
                    {
                        var charge = 1;
                        var isoFit = 0;
                        var mass = clsUtilities.ConvoluteMass(spectrum.IonsMZ[ionIndex], 1, 0);
                        var peakFWHM = 0;
                        var signalToNoise = spectrum.IonsIntensity[ionIndex] / baselineNoiseLevel;
                        var monoisotopicAbu = -10;
                        var monoPlus2Abu = -10;

                        mBPIWriter.WriteDecon2LSIsosFileEntry(
                            dataWriter, scanNumber, charge,
                            spectrum.IonsIntensity[ionIndex], spectrum.IonsMZ[ionIndex],
                            isoFit, mass, mass, mass,
                            peakFWHM, signalToNoise, monoisotopicAbu, monoPlus2Abu);
                    }
                }
            }
        }

        private void SavePEKFileToDiskWork(
            TextWriter writer,
            clsScanInfo currentScan,
            clsSpectraCache spectraCache,
            string inputFileName,
            bool fragmentationScan,
            ref int spectrumExportCount)
        {
            var exportCount = 0;

            if (!spectraCache.GetSpectrum(currentScan.ScanNumber, out var spectrum, true))
            {
                SetLocalErrorCode(clsMASIC.eMasicErrorCodes.ErrorUncachingSpectrum);
                return;
            }

            spectrumExportCount += 1;

            writer.WriteLine("Time domain signal level:" + "\t" + currentScan.BasePeakIonIntensity.ToString("0.000"));          // Store the base peak ion intensity as the time domain signal level value

            writer.WriteLine("MASIC " + mOptions.MASICVersion);                     // Software version
            var dataLine = "MS/MS-based PEK file";
            if (mOptions.RawDataExportOptions.IncludeMSMS)
            {
                dataLine += " (includes both survey scans and fragmentation spectra)";
            }
            else
            {
                dataLine += " (includes only survey scans)";
            }

            writer.WriteLine(dataLine);

            int scanNumber;
            if (mOptions.RawDataExportOptions.RenumberScans)
            {
                scanNumber = spectrumExportCount;
            }
            else
            {
                scanNumber = currentScan.ScanNumber;
            }

            dataLine = "Filename: " + inputFileName + "." + scanNumber.ToString("00000");
            writer.WriteLine(dataLine);

            if (fragmentationScan)
            {
                writer.WriteLine("ScanType: Fragmentation Scan");
            }
            else
            {
                writer.WriteLine("ScanType: Survey Scan");
            }

            writer.WriteLine("Charge state mass transform results:");
            writer.WriteLine("First CS,    Number of CS,   Abundance,   Mass,   Standard deviation");

            if (spectrum.IonCount > 0)
            {
                // Populate intensities and pointerArray()

                var intensities = new double[spectrum.IonCount];
                var pointerArray = new int[spectrum.IonCount];
                for (var ionIndex = 0; ionIndex < spectrum.IonCount; ionIndex++)
                {
                    intensities[ionIndex] = spectrum.IonsIntensity[ionIndex];
                    pointerArray[ionIndex] = ionIndex;
                }

                // Sort pointerArray() based on the intensities in intensities
                Array.Sort(intensities, pointerArray);

                int startIndex;

                if (mOptions.RawDataExportOptions.MaxIonCountPerScan > 0)
                {
                    // Possibly limit the number of ions to maxIonCount
                    startIndex = spectrum.IonCount - mOptions.RawDataExportOptions.MaxIonCountPerScan;
                    if (startIndex < 0)
                        startIndex = 0;
                }
                else
                {
                    startIndex = 0;
                }

                // Define the minimum data point intensity value
                var minimumIntensityCurrentScan = spectrum.IonsIntensity[pointerArray[startIndex]];

                // Update the minimum intensity if a higher minimum intensity is defined in .IntensityMinimum
                minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, mOptions.RawDataExportOptions.IntensityMinimum);

                // If mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio is > 0, then possibly update minimumIntensityCurrentScan
                if (mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio > 0)
                {
                    minimumIntensityCurrentScan = Math.Max(minimumIntensityCurrentScan, currentScan.BaselineNoiseStats.NoiseLevel * mOptions.RawDataExportOptions.MinimumSignalToNoiseRatio);
                }

                exportCount = 0;
                for (var ionIndex = 0; ionIndex < spectrum.IonCount; ionIndex++)
                {
                    if (spectrum.IonsIntensity[ionIndex] >= minimumIntensityCurrentScan)
                    {
                        var dataLine1 =
                            "1" + "\t" +
                            "1" + "\t" +
                            spectrum.IonsIntensity[ionIndex] + "\t" +
                            spectrum.IonsMZ[ionIndex] + "\t" +
                            "0";

                        writer.WriteLine(dataLine1);
                        exportCount += 1;
                    }
                }
            }

            writer.WriteLine("Number of peaks in spectrum = " + spectrum.IonCount.ToString());
            writer.WriteLine("Number of isotopic distributions detected = " + exportCount.ToString());
            writer.WriteLine();
        }

        private void SaveRawDataToDiskWork(
            StreamWriter dataWriter,
            StreamWriter scanInfoWriter,
            clsScanInfo currentScan,
            clsSpectraCache spectraCache,
            string inputFileName,
            bool fragmentationScan,
            ref int spectrumExportCount)
        {
            switch (mOptions.RawDataExportOptions.FileFormat)
            {
                case clsRawDataExportOptions.eExportRawDataFileFormatConstants.PEKFile:
                    SavePEKFileToDiskWork(dataWriter, currentScan, spectraCache, inputFileName, fragmentationScan, ref spectrumExportCount);
                    break;

                case clsRawDataExportOptions.eExportRawDataFileFormatConstants.CSVFile:
                    SaveCSVFilesToDiskWork(dataWriter, scanInfoWriter, currentScan, spectraCache, fragmentationScan, ref spectrumExportCount);
                    break;

                default:
                    // Unknown format
                    // This code should never be reached
                    break;
            }
        }
    }
}
